Rebuild docs (#86)
ChainSafe/blst-ts: doc/values.md

# JavaScript Values Under-the-Hood

> _**note:** If you followed the development of this file you may have noticed that it took a very long time to write and changed during writing.  There is a lot of inconsistency in the writings surrounding [`HandleScope`](https://github.com/nodejs/node-addon-api/blob/main/doc/handle_scope.md)._ _I built the code using what I read so it "works correctly" but at the time I had no idea why, nor how._ _To try an explain it better I read what was available and wrote about it._ _I realized it didn't "make sense."_ _So I read through the `node` source to try and figure it out but there is a TON of code that makes "values work."_ _I revised this document a couple times during the process as my understanding evolved._ _I think this is how it works but have [requested a review](https://github.com/nodejs/node/issues/47504) from the V8 team to make sure its correct._ _I also want to hightlight this body of work in hopes it, or parts of it, may be added to the [official documentation](https://github.com/nodejs/node-addon-api/tree/main/doc) or the [examples repo](https://github.com/nodejs/node-addon-examples) to move the content up the open-source tree where it will be more visible to the broader community._

Reviewer: ```suggestion
_**note:** If you followed the development of this file you may have noticed that it took a very long time to write and changed during writing.  There is a lot of inconsistency in the writings surrounding [`HandleScope`](https://github.com/nodejs/node-addon-api/blob/main/doc/handle_scope.md)._ _I built the code using what I read so it "works correctly" but at the time I had no idea why, nor how._ _To try and explain it better I read what was available and wrote about it._ _I realized it didn't "make sense"._ _So I read through the `node` source to try and figure it out but there is a TON of code that makes "values work."_ _I revised this document a couple times during the process as my understanding evolved._ _I think this is how it works but have [requested a review](https://github.com/nodejs/node/issues/47504) from the V8 team to make sure its correct._ _I also want to highlight this body of work in hopes it, or parts of it, may be added to the [official documentation](https://github.com/nodejs/node-addon-api/tree/main/doc) or the [examples repo](https://github.com/nodejs/node-addon-examples) to move the content up the open-source tree where it will be more visible to the broader community._
```

Author: Updated.

Because data is heap allocated, values provide two functions.  They facilitate lookup and garbage collection and there are two different systems for accomplishing that.  `v8::Context` is what [creates](./environment.md#) lexical context.  `internal::v8::Object` [doclink](https://github.com/nodejs/node/blob/4166d40d0873b6d8a0c7291872c8d20dc680b1d7/deps/v8/src/objects/objects.h#L299) and handles to those objects allow the runtime to function with data that moves.

You can find the full inheritance hierarchy [here](./reference.md#object-inheritance-hierarchy), but pay attention to the following section as it will help to understand how `HandleScope` and `napi_value`  relate.

```c++
// - Object
//   - Smi          (immediate small integer)
//   - TaggedIndex  (properly sign-extended immediate small integer)
//   - HeapObject   (superclass for everything allocated in the heap)
//     - JSReceiver  (suitable for property access) ie. Object/Proxy
//     - FixedArrayBase  (superclass for BufferLike and hash-table based data)
//     - PrimitiveHeapObject  (primitives String/Symbol/Number)
```

> The `Isolate` is responsible for handling all data related to JS execution and values are only the slice that addon developers interact with.  Memory is allocated in the heap to manage all things JS and native code handles  are references, to pointers, to said data.

Reviewer 2: ```suggestion
The `Isolate` is responsible for handling all data related to JS execution and values are only the slice that addon developers interact with.  Memory is allocated in the heap to manage all things JS, and native code handles are references, to pointers, to said data.
```

Author: Updated.

When working with `napi_value`, and by extension `Napi::Value`, it is important to pass by value.  While it is possible to also pass by reference, which is glorified pass by value, one should never pass by pointer. That insinuates some form of heap allocation for the `napi_value` and the value is not properly counted in the references because of the extra level of indirection.

It is important to note that `v8::Object` and a JavaScript `Object` are not the same. In a JS script, `new Object` creates an instance of `JSObject` under-the-hood and that is a subclass of `JSReceiver`.

> The isolate creates objects during runtime and they will be stored in the faster to access space. The runtime will retain a `Handle` to the data so values can be looked up. The `Handle` will be added to, and managed by, the nearest `HandleScope` so that the GC can cleanup once the `HandleScope` is destroyed (variable goes out of scope lexically).

Reviewer 2: ```suggestion
The isolate creates objects during runtime and they will be stored in the faster-to-access space. The runtime will retain a `Handle` to the data so values can be looked up. The `Handle` will be added to, and managed by, the nearest `HandleScope` so that the GC can cleanup once the `HandleScope` is destroyed (variable goes out of scope lexically).
```

Author: Updated.

> When the GC runs over the fast access space, any objects that are still relevant are [assigned storage](https://github.com/nodejs/node/blob/4166d40d0873b6d8a0c7291872c8d20dc680b1d7/deps/v8/src/objects/objects.h#L378) by the `Isolate` in the slower to access space.

Reviewer 2: ```suggestion
When the GC runs over the fast-access space, any objects that are still relevant are [assigned storage](https://github.com/nodejs/node/blob/4166d40d0873b6d8a0c7291872c8d20dc680b1d7/deps/v8/src/objects/objects.h#L378) by the `Isolate` in the slower-to-access space.
```

Author: Updated.

> Values, `napi_value` for the rest of this document will be referred to as Value, and similarly a `v8::Object` will also be referred to simply as Object, will remain in scope for as long as the native function is on the stack.  When the garbage collector runs (at some point, but assume) while a function is somewhere on the stack, all referenced objects on the stack that are not already moved, will be moved with longer-lived variables. This is why pointing to the underlying data directly is not a great idea and patterns explained below were developed.

Reviewer 2: ```suggestion
A value, `napi_value` for the rest of this document will be referred to as Value, and similarly a `v8::Object` will also be referred to simply as Object, will remain in scope for as long as the native function is on the stack.  When the garbage collector runs (at some point, but assume) while a function is somewhere on the stack, all referenced objects on the stack that are not already moved, will be moved with longer-lived variables. This is why pointing to the underlying data directly is not a great idea and patterns explained below were developed.
```

Author: Updated.

  /// Casts to another type of `Napi::Value`, when the actual type is known or
  /// assumed.
  ///
  /// This conversion does NOT coerce the type. Calling any methods
  /// inappropriate for the actual value type will throw `Napi::Error`.
  template <typename T>
  T As() const;

  MaybeOrValue<Boolean> ToBoolean() const;  ///< Coerces a value to a JavaScript boolean.
  MaybeOrValue<Number> ToNumber() const;  ///< Coerces a value to a JavaScript number.
  MaybeOrValue<String> ToString() const;  ///< Coerces a value to a JavaScript string.
  MaybeOrValue<Object> ToObject() const;  ///< Coerces a value to a JavaScript object.

 protected:
  napi_env _env;
  napi_value _value;
};
```

So back to the part about requesting values. In order to access JavaScript values, the executing code must be on the JS thread. A challenge presents itself because one of the most important reasons for writing node addons is to circumvent the limitations of strictly staying on thread. When working with the JS data off thread there are provisions that need to be made so that the data is accessible.

## How to Use the Data Backing Values

> There are a few ways to do this with the simplest being copy it.  A quick `memcpy` or assignment to another variable, while on the thread, will allow that data to be used after leaving the JS context.  One simply asks `v8` for the value, saves it to a class member somewhere and off we go.  That is totally valid for some cases, but in others it's SLOW...

Reviewer 2: ```suggestion
There are a few ways to do this with the simplest being to copy it.  A quick `memcpy` or assignment to another variable, while on the thread, will allow that data to be used after leaving the JS context.  One simply asks `v8` for the value, saves it to a class member somewhere and off we go.  That is totally valid for some cases, but in others it's SLOW...
```

Author: Updated.
